rust-runtime/aws-smithy-http-server/src/exception.rs

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Exception types.
+//!
+//! This module contains "exception types". There is only one such type, [`SmithyFrameworkException`].
+//!
+//! As opposed to rejection types (see [`crate::rejection`]), which are an internal detail about
+//! the framework, exceptions are surfaced to clients in HTTP responses: indeed, they implement
+//! [`axum_core::response::IntoResponse`]. Rejections can be "grouped" and converted into an
+//! exception type: for example, all request rejections due to serialization issues can be
+//! conflated under the [`SmithyFrameworkExceptionType::Serialization`] enum variant.
+//!
+//! The HTTP response representation of an exception can be protocol-specific: for example,
+//! exceptions in the RestJson1 protocol set the `X-Amzn-Errortype` header.
+//!
+//! Exception types' behavior doesn't have anything to do with how _programming language exceptions_
+//! behave: there are no exceptions in Rust, `[SmithyFrameworkException`] is just a regular Rust
+//! enum. Their name comes from their representation in all AWS protocols: for example,
+//! [`SmithyFrameworkExceptionType::Serialization`] renders an `X-Amzn-Errortype` header with the
+//! value `SerializationException` when using the RestJson1 protocol.
+//!
+//! Generated code works always works with [`crate::rejection`] types when deserializing requests
+//! and serializing response. Just before a response needs to be sent, the generated code looks up
+//! and converts into the corresponding exception type, and then it uses the exception's
+//! [`axum_core::response::IntoResponse`] implementation to render and send a response.
+
+use crate::protocols::Protocol;
+
+#[derive(Debug)]
+pub enum SmithyFrameworkExceptionType {
+    // UnknownOperation,
+    Serialization(crate::Error),
+    // UnsupportedMediaType,
+    // NotAcceptable,
+}
+
+/// String representation of the exception type.
+/// Used as the value of the `X-Amzn-Errortype` header in RestJson1.
+/// Used as the value passed to construct an [`crate::ExtensionRejection`].
+impl SmithyFrameworkExceptionType {
+    pub fn name(&self) -> &'static str {
+        match self {
+            SmithyFrameworkExceptionType::Serialization(_) => "SerializationException",
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SmithyFrameworkException {
```

Comment:
I'm tempted to rename this, it might not be a good name. I don't like putting in the word "framework", and the concept of exceptions is not inherent to Smithy, it's just a side-product of how AWS protocols serialize errors in responses.

What if we just name it `Exception`?

Reply:
I am ok naming this `Exception`.. it can still be very confusing for many people as it will be associated with the concept of exception in java / python / etc.. I believe we just need to document this very clearly and we can rename it to just Exception.

Reply:
An alternative is to come up with a name of our own, different than exception, rejection, or error. So, name this type `X` and make the fact that `X::Z` in responses gets written as `ZException` just a detail of the protocol. After all, the fact that the word "exception" has been used in all AWS protocol errors might does not imply that it will continue to do so. Perhaps in the future a protocol we want to support (AWS or not) serializes an error as "Mistake", so maybe it's better to come up with our own name.

Reply:
The more I think about it, the more I want to get far away from `Exception` for the reason you provided above.. 

I think it is better if we come up with our name. In my mind this is an error, hence it should be suffixed with `Error`. What about `SmithyRuntimeError` or just `RuntimeError` ?

Reply:
We think coming up with a new name is better. For now we'll go with `RuntimeError`.